File: java/src/Leetcode566.java
public class Leetcode566{
    public int[][] matrixReshape(int[][] mat, int r, int c) {
        int m = mat.length, n = mat[0].length;
        if(m * n != r * c)
            return mat;
        
        int[][] result = new int[r][c];
        int rowIndex = 0, colIndex = 0;
        for(int i = 0; i < r; i++){
            int[] row = new int[c];
            for(int j = 0; j < c; j++){
                if(colIndex < n){
                    row[j] = mat[rowIndex][colIndex++];
                }else{
                    colIndex = 0;
                    rowIndex++;
                    row[j] = mat[rowIndex][colIndex++];
                }
            }
            result[i] = row;
        }
        return result;
    }
}

File: python/Leetcode566.py
from typing import List

class Leetcode566:
    def matrix_reshape(self, mat: List[List[int]], r: int, c: int) -> List[List[int]]:
        entries = len(mat) * len(mat[0])
        if entries != r * c:
            return mat
        m = len(mat)
        n = len(mat[0])
        r_idx = 0 # mat row
        c_idx = 0 # mat column
        result = []
        for i in range(r):
            row = []
            for j in range(c):
                if c_idx < n:
                    row.append(mat[r_idx][c_idx])
                    c_idx += 1
                else:
                    r_idx += 1
                    c_idx = 0
                    row.append(mat[r_idx][c_idx])
                    c_idx += 1
            result.append(row)
        return result